### PR TITLE
Algo Benchmark Library card catalog

### PR DESCRIPTION
**Pull request by mpharrigan**
This "card catalog" will be the centralized record of algorithmic benchmarks that are highly organized and use the new `cirq_google.workflow` cirqflow style of execution.

This PR does not commit any benchmarks, but establishes (via the dataclasses and the tests) a set of conventions for organizing things.

### CODE REVIEW

File: recirq/algo_benchmark_library.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+from functools import lru_cache
+from typing import Type, Callable, List, Tuple
+
+import pandas as pd
+
+from cirq_google.workflow import ExecutableSpec, QuantumExecutableGroup
+
+
+@dataclass
+class AlgoBenchmark:
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: done 77e48f956073c36437579ec89a97a2d81cca4fab

File: recirq/algo_benchmark_library.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    domain: str
+    """The problem domain. Usually a high-level ReCirq module."""
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: done 9481205713eb4eacd8a0b5555bc5c0215e05ac2b

[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: done 7c7faa1a760606cc3dee362124ab00af689a224d

File: recirq/algo_benchmark_library.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    for this AlgoBenchmark lives.
+    """
+
+    spec_class: Type[ExecutableSpec]
+    """The ExecutableSpec subclass for this AlgoBenchmark."""
+
+    data_class: Type
+    """The class which can contain ETL-ed data for this AlgoBenchmark."""
+
+    gen_func: Callable[[...], QuantumExecutableGroup]
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: done 6f9224c5171c9b6e5ddf98b771228e879df9ee6d

File: recirq/algo_benchmark_library.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+@dataclass
+class BenchmarkConfig:
+    short_name: str
+    """The short name for this config. Unique within an AlgoBenchmark."""
+
+    full_name: str
+    """A globally unique name for this config."""
+
+    gen_script: str
+    """The script filename that generates the QuantumExecutableGroup for this Config."""
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: done 69b6b65ab87bb49827ea5fce392e8e2d215358ae

File: recirq/algo_benchmark_library.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    run_scripts: List[str]
+    """A list of script filenames that execute (or can execute) this config."""
+
+
+BENCHMARKS = [
+]
+
+
+@lru_cache()
+def get_all_algo_configs() -> List[Tuple[AlgoBenchmark, BenchmarkConfig]]:
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: done 8f139ecab5a1ba22a2bd9ed92833a2e1b9ba50f6

File: recirq/algo_benchmark_library.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    ret = []
+    for algo in BENCHMARKS:
+        for config in algo.configs:
+            ret.append((algo, config))
+    return ret
+
+
+def main():
+    df = pd.DataFrame([algo.as_strings() for algo in BENCHMARKS]).set_index('executable_family')
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library.py]: sortof forgot I left this in here. Changed the name of this function and removed the `__main__` entry point 1514b2069fe447e4437a0968b74dc54b6e617312

The nice thing about pandas is you can use the same technique in a juypter noteook and get a nice html table that you can paste into sheets or a doc

File: recirq/algo_benchmark_library_test.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+def test_benchmark_name_unique_in_domain():
+    # In a given domain, all benchmark names should be unique
+    pairs = [(algo.domain, algo.name) for algo in BENCHMARKS]
+    assert len(set(pairs)) == len(pairs)
+
+
+@pytest.mark.parametrize('algo', BENCHMARKS)
+def test_executable_family_is_formulaic(algo):
+    # Check consistency in the AlgoBenchmark dataclass:
+    assert algo.executable_family == algo.spec_class.executable_family
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply recirq/algo_benchmark_library_test.py]: done 3308c5e2fdff9b87e6c49d2b27863d630968525b

**Review by mpharrigan — COMMENTED** [review]
ptal

File: recirq/algorithmic_benchmark_library.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+import dataclasses
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import Type, Callable, List, Tuple
+
+import pandas as pd
+
+from cirq_google.workflow import ExecutableSpec, QuantumExecutableGroup
```
Comment (dstrain115) [inline recirq/algorithmic_benchmark_library.py:8]: From the pytest:

`ModuleNotFoundError: No module named 'cirq_google.workflow'`

Is this waiting on another PR?

**Review by dstrain115 — APPROVED** [review]
LGTM once the tests resolve.

File: recirq/algorithmic_benchmark_library.py (hunk 8 above)
Comment (mpharrigan) [reply recirq/algorithmic_benchmark_library.py:8]: This and other forthcoming code relies on the currently-unreleased `cirq_google.workflow` module. I'll have to figure out a way to make this work in the mean time. 

1. release openfermion that doesn't pin to a particular cirq version
2. enable testing against multiple versions of cirq including prerelease in ReCirq CI
3. put import guards and conditional testing so these tests only run if a recent enough cirq version is available

Comment (mpharrigan) [reply recirq/algorithmic_benchmark_library.py]: Following #237 , this is now implemented. The tests only get run on the `next` config until a release of cirq is cut.

**Pull request merged**